### PR TITLE
[Feature](mlu-ops):perf voxelization.

### PR DESCRIPTION
**Pull request by mahxn0**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation
performance optimization 

## 2. Modification

modify: bangc-ops/kernels/voxelization/voxelization_kernel.mlu
            docs/bangc-docs/design_docs/voxelization/voxelization.md
add:  docs/bangc-docs/design_docs/voxelization/mask.png
         docs/bangc-docs/design_docs/voxelization/gather.jpg
         docs/bangc-docs/design_docs/voxelization/gather_scatter.jpg
## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- static threshold
  - diff1
    - [ ] float32 mlu diff1 <= 1e-5
    - [ ] float32 mlu diff1 <= 3e-3
    - [ ] float16 mlu diff1 <= 3e-3
  - diff2
    - [ ] float32 mlu diff2 <= 1e-5
    - [ ] float32 mlu diff2 <= 3e-3
    - [ ] float16 mlu diff2 <= 3e-3
  - diff3
    - [x] mlu diff3 == 0
    - [ ] mlu diff3_1 == 0
    - [ ] mlu diff3_2 == 0
- dynamic threshold
  - [ ] diff1: mlu diff1 <= max(baseline diff1 * 10, static threshold)
  - [ ] diff2: mlu diff2 <= max(baseline diff2 * 10, static threshold)
  - [ ] diff3: mlu diff3 <= max(baseline diff3 * 10, static threshold)
    - float32, threshold = 1e-5
    - float16, threshold = 1e-3

#### 3.1.2 Operator Scheme checklist

- Supported hardware
  - [x] MLU370
  - [x] MLU590
- Job types
  - [x] BLOCK
  - [x] UNION1
  - [ ] UNION2
  - [ ] UNION4
  - [ ] The operator will dynamically select the most suitable task type, for example, UNION8

#### 3.2.1 Accuracy Test

If you have checked the following items, please tick the relevant box.

- [x] Data type test (e.g. float32/int8)
- [x] Multi-dimensional tensor test
- [ ] Layout test
- [x] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [x] stability test
- [x] Multiple platform test
- [ ] Gen_case module test, see: [Gencase-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/Gencase-User-Guide-zh.md)
- [x] Nan/INF tests 
- [ ] Bug fix tests
- [x] For memory leak check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [x] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md)
- [x] For I/O calculation efficiency check details, see: [MLU-OPS-Performance-Acceptance-Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md)

#### 3.2.2 Parameter Check

Test Point-1: `When a new operator is submitted, the test points are given and the test results are stated`. Acceptance Standard: `Normal error`.
```bash
Please fill your test results(Error Message) in here, ...
```

Test Point-2: `Whether illegal parameters are passed`. Acceptance Standard: `Normal error`.
```bash
Test results...
```

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform compare between baseline and perf on MLU590

shape | baseline time | every kernel time | memory perf time | every kernel time | kernel perf pencent | total pencent | memory vector perf time | every kernel time | kernel perf pencent | total pencent
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
{'input_dim': [[53429, 237], [3], [6]]} | 43410 | KernelDynamicVoxelize: 265KernelPoint2Voxel:1435KernelCalcPointsPerVoxel:41212KernelAssignVoxelsCoors: 498 | 15964 | KernelDynamicVoxelize: 265KernelPoint2Voxel:1435KernelCalcPointsPerVoxel:14182KernelAssignVoxelsCoors: 82 | 00290.6%607% | 271.9% | 2960 | KernelDynamicVoxelize: 265KernelPoint2Voxel:1435KernelCalcPointsPerVoxel:1178KernelAssignVoxelsCoors: 82 | 003498%607% | 1466%
{'input_dim': [[8800, 9], [3], [6]]} | 7163 | KernelDynamicVoxelize: 17KernelPoint2Voxel:270KernelCalcPointsPerVoxel: 6773KernelAssignVoxelsCoors: 103 | 767 | KernelDynamicVoxelize: 17KernelPoint2Voxel:270KernelCalcPointsPerVoxel: 436KernelAssignVoxelsCoors: 44 | 001553%234% | 933% | 488 | KernelDynamicVoxelize: 17KernelPoint2Voxel:270KernelCalcPointsPerVoxel:157KernelAssignVoxelsCoors: 44 | 004314%234% | 1467%

### 3.4 Summary Analysis
Mainly use gather/scatter to optimize the discrete visit memory, quantify the calculation process, and improve performance as expected，
Follow-up consideration to further optimize with sram and u1 tasks
.

### CODE REVIEW

File: docs/bangc-docs/design_docs/voxelization/voxelization.md [hunk 1]
```diff
@@ -387,6 +388,58 @@ for (int32_t p_idx = points_start; p_idx < points_end; ++p_idx) {
...
+
+point_to_voxelidx: 当前points为voxel中的第几个point，其中0的数量代表num_voxel的数量．
+
+假设这里第一个kernel输出的point的体素坐标为：[1, 2, 3, 3, 2, 5, 6, 6, 2, 2, 10, 11, 12, 13]
+
+则第二个kernel输出的
+
+point_to_pointidx为：[0, 1, 2, 2, 1, 5, 6, 6, 1, 1, 10, 11, 12, 13]
+
+point_to_voxelidx为：[0, 0, 0, 1, 1, 0, 0, 1, 1, 2, 0, 0, 0, 0]
```
Comment (PetrelYy) [inline docs/bangc-docs/design_docs/voxelization/voxelization.md:407]: gather.png 这个图字糊了

File: docs/bangc-docs/design_docs/voxelization/voxelization.md [hunk 2]
```diff
@@ -387,6 +388,58 @@ for (int32_t p_idx = points_start; p_idx < points_end; ++p_idx) {
...
+
+第一，先计算每个voxel中的第一个点所属voxel_idx．
+
+第二，计算每个voxel中重复的点所属的voxel_idx（重复的点的voxel_idx取第一个point所属的voxel_idx）.
+
+具体为：计算core上一次能处理的最大points数量max_points_count和循环次数repeat．
+
+- step1：先计算每个voxel中的第一个点所属voxel_idx
+
+  - 统计当前循环中p2v_idx中0的数量，即体素中第一个元素的数量，也代表体素的数量．
```
Comment (PetrelYy) [inline docs/bangc-docs/design_docs/voxelization/voxelization.md:421]: 整体思路没啥问题，表述和标点错用统一再check下

**Review by DanieeelLiu — APPROVED** [review]

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu [hunk 3]
```diff
@@ -470,28 +470,211 @@ __mlu_global__ void mluCalcPointsPerVoxel(
...
+
+      // step1: load p2p/p2v data.
+      __memcpy_async(nram_p2p_idx, point_to_pointidx + points_offset,
+                     deal_num * sizeof(int32_t), GDRAM2NRAM);
+      __memcpy(nram_p2v_idx, point_to_voxelidx + points_offset,
+               deal_num * sizeof(int32_t), GDRAM2NRAM);
+
+      // step2:　process the first point in voxel
+      __bang_eq_scalar(nram_temp_mask, nram_p2v_idx, 0, deal_num);
+      __bang_write_value(nram_scatter_output, deal_num, -1);
```
Comment (Unireverse) [inline bangc-ops/kernels/voxelization/voxelization_kernel.mlu:528]: 这里是不是可以用move流从sram上搬0，这样可以和计算并行

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu [hunk 4]
```diff
@@ -470,28 +470,211 @@ __mlu_global__ void mluCalcPointsPerVoxel(
...
+      __bang_mul(nram_scatter_offset, nram_temp_mask, nram_base_offset,
```
Comment (Unireverse) [inline bangc-ops/kernels/voxelization/voxelization_kernel.mlu:529]: 另外，这里nram_scatter_offset作为dst，528行有啥意义吗

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu [hunk 5]
```diff
@@ -470,28 +470,211 @@ __mlu_global__ void mluCalcPointsPerVoxel(
...
+                 deal_num);
+      int32_t cur_loop_num_voxels =
+          __bang_filter((float *)nram_scatter_offset, (float *)nram_base_offset,
+                        (float *)nram_temp_mask, deal_num);
+      __bang_write_value(nram_scatter_mask, deal_num, 1);
+      int32_t reserve_voxels = cur_loop_num_voxels;
+      int32_t vir_total_num_voxels = voxel_num_temp + cur_loop_num_voxels;
+      if (vir_total_num_voxels >= max_voxels) {
+        reserve_voxels = max_voxels - voxel_num_temp;
+        __bang_write_value(nram_scatter_mask + reserve_voxels, deal_num, 0);
```
Comment (Unireverse) [inline bangc-ops/kernels/voxelization/voxelization_kernel.mlu:539]: 这里写0的数量是不是有问题

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu [hunk 6]
```diff
@@ -470,28 +470,211 @@ __mlu_global__ void mluCalcPointsPerVoxel(
...
+                  sizeof(int32_t), NRAM2NRAM, sizeof(int32_t), reserve_voxels);
+        __memcpy(num_points_per_voxel + voxel_num_temp, nram_scatter_mask,
+                 reserve_voxels * sizeof(int32_t), NRAM2GDRAM);
+        voxel_num_temp += reserve_voxels;
+      }
+      // store voxel first points.
+      __memcpy(coor_to_voxelidx + points_offset, nram_scatter_output,
+               deal_num * sizeof(int32_t), NRAM2GDRAM);
+
+      // step2: process other points in voxels.
```
Comment (Unireverse) [inline bangc-ops/kernels/voxelization/voxelization_kernel.mlu:561]: step2？

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu [hunk 7]
```diff
@@ -470,28 +470,211 @@ __mlu_global__ void mluCalcPointsPerVoxel(
...
+      float *gather_mask = (float *)nram_p2v_idx;
+      int32_t *gather_output = (int32_t *)nram_scatter_output;
+      __bang_gt(gather_mask, (float *)nram_p2v_idx, (float *)nram_temp_zeros,
+                deal_num);
+      const int32_t count = __bang_filter(
+          (float *)nram_p2p_idx, (float *)nram_p2p_idx, gather_mask, deal_num);
+      if (count > 0) {
+        __bang_mul_scalar(nram_p2p_idx, nram_p2p_idx, sizeof(int32_t), count);
+        __gather(gather_output, coor_to_voxelidx, (unsigned int *)nram_p2p_idx,
```
Comment (Unireverse) [inline bangc-ops/kernels/voxelization/voxelization_kernel.mlu:570]: 片上是move流，不知道有没有sync风险，找人问问

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu [hunk 8]
```diff
@@ -470,28 +470,211 @@ __mlu_global__ void mluCalcPointsPerVoxel(
...
+        __bang_gt_bitindex((float *)nram_mask_bitindex,
+                           (float *)nram_scatter_mask, (float *)nram_temp_zeros,
+                           PAD_UP(count, 8));
+        __bang_add_scalar(nram_temp_mask, nram_base_offset, points_offset,
+                          deal_num);
+        __bang_filter((float *)nram_temp_mask, (float *)nram_temp_mask,
+                      gather_mask, deal_num);
+        __bang_mul_scalar(nram_temp_mask, nram_temp_mask, sizeof(int32_t),
+                          deal_num);
+        __scatter(coor_to_voxelidx, gather_output,
```
Comment (Unireverse) [inline bangc-ops/kernels/voxelization/voxelization_kernel.mlu:583]: 同上

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu (hunk 3 above)
Comment (mahxn0) [reply bangc-ops/kernels/voxelization/voxelization_kernel.mlu:528]: 这里只有一条比较指令，用sram收益不大．

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu (hunk 4 above)
Comment (mahxn0) [reply bangc-ops/kernels/voxelization/voxelization_kernel.mlu:529]: 默认将不满足条件的voxel_id都置为-1

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu (hunk 5 above)
Comment (mahxn0) [reply bangc-ops/kernels/voxelization/voxelization_kernel.mlu:539]: 正确来说是reserve_voxels，但是这里reserve_voxels可能会是0，不加判断指令会报错．
写deal_num是因为nram_scatter_mask后面一块内存当前并没有用到，为了代码简洁就这样写了

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu (hunk 6 above)
Comment (mahxn0) [reply bangc-ops/kernels/voxelization/voxelization_kernel.mlu:561]: 已修改

File: bangc-ops/kernels/voxelization/voxelization_kernel.mlu (hunk 7 above)
Comment (mahxn0) [reply bangc-ops/kernels/voxelization/voxelization_kernel.mlu:581]: 这里是gdram2nram是io流　我理解没有风险的哈．

**Review by duzekunKTH — APPROVED** [review]

**Pull request merged**